src/workflows/ingestion/demux/main.nf

```diff
@@ -1,125 +1,60 @@
-nextflow.enable.dsl=2
-
-workflowDir = params.rootDir + "/src/workflows"
-targetDir = params.rootDir + "/target/nextflow"
-
-include { cellranger_mkfastq } from targetDir + "/demux/cellranger_mkfastq/main.nf"
-include { bcl_convert } from targetDir + "/demux/bcl_convert/main.nf"
-include { bcl2fastq } from targetDir + "/demux/bcl2fastq/main.nf"
-include { fastqc } from targetDir + "/qc/fastqc/main.nf"
-include { multiqc } from targetDir + "/qc/multiqc/main.nf"
-
-include { readConfig; channelFromParams; preprocessInputs; helpMessage } from workflowDir + "/utils/WorkflowHelper.nf"
-include { passthroughMap as pmap } from workflowDir + "/utils/DataflowHelper.nf"
-
-config = readConfig("$workflowDir/ingestion/demux/config.vsh.yaml")
-
-workflow demux {
-  helpMessage(config)
-
-  channelFromParams(params, config)
-    | run_wf
-}
-
 workflow run_wf {
   take:
   input_ch
 
   main:
-  commonOptions = [
-    args: [ output: "fastq/\$id" ],
-    auto: [ publish: true ]
-  ]
-  preprocessed_ch = input_ch
-    | preprocessInputs("config": config)
-
-  mkfastq_ch = preprocessed_ch
-    | filter{ it[1].demultiplexer == "mkfastq" }
-    | cellranger_mkfastq.run(commonOptions)
-    
-  bcl_convert_ch = preprocessed_ch
-    | filter{ it[1].demultiplexer  == "bclconvert" }
-    | bcl_convert.run(commonOptions)
-
-  bcl2fastq_ch = preprocessed_ch
-    | filter{ it[1].demultiplexer  == "bcl2fastq" }
-    | bcl2fastq.run(commonOptions)
-
-  /* Combine the different demultiplexer channels */
-  all_ch =
-    mkfastq_ch
-      | mix( bcl_convert_ch, bcl2fastq_ch )
-      | map {  tup ->
-        [tup[0], tup[1].output]
-      }
+  output_ch = input_ch
+
+    // translate the demultiplexer argument to the component name that needs to be run
+    | map { id, state ->
+      def funcNameMapper = [
+        bclconvert: "bcl_convert",
+        bcl2fastq: "bcl2fastq",
+        mkfastq: "cellranger_mkfastq"
+      ]
+      def funcName = funcNameMapper[state.demultiplexer]
+      def newState = state + [funcName: funcName]
+      [id, newState]
+    }
 
-  /* Generate fastqc reports for every sample */
-  all_ch
-    | fastqc.run(
-        [
-          args: [ mode: "dir", output: "fastqc/\$id" ],
-          auto: [ publish: true ]
+    // run the demultiplexers
+    | runComponents(
```

Comment:
Todo: update to `runEach`.

Note that `{ id, state, config ->` will need to be updated to `{ id, state, component`, and `config.functionality` to `component.config.functionality`.

Reply:
See 2a703b5